docs/design/diagnostics.md

```diff
@@ -0,0 +1,293 @@
+func NewWarningDiagnostic(summary string, detail string, attribute *tftypes.AttributePath) Diagnostic
+```
+
+This type of implementation will still be difficult to extend or change.
+
+As another extensibility example, as `Diagnostics` pass through various functions across subsystems, determining the "source" or "cause" of a diagnostic must be done by `Summary` and/or `Detail` value similar to checking Go `error` types by value using `(error).Is()`. Extending the `Diagnostic` type is possible to add other identifying fields or creating new types, e.g.
+
+```go
+type ValidationDiagnostic struct {
+  Diagnostic
+}
+```
+
+However, the implementation for diagnostics then differs from other design choices in the framework where interfaces are preferred. For example, it becomes difficult to implement equality and future enhancements for diagnostics since the interface is not strongly defined for extensions.
+
+### Basic Interface
+
+Diagnostics could be modeled around interfaces:
+
+```go
+package diag
+
+type Diagnostic interface {
+  Severity() Severity
+  Summary()  string
+  Detail()   string
+
+  Equal(other Diagnostic) bool
+}
+
+type DiagnosticWithLogger interface {
+  Diagnostic
+
+  Log(ctx context.Context)
+}
+
+// This could be folded into Diagnostic
+type DiagnosticWithPath interface {
+  Diagnostic
+
+  Path() *tftypes.AttributePath // See also: https://github.com/hashicorp/terraform-plugin-framework/issues/81
+}
+
+type Diagnostics []Diagnostic
+
+func (d Diagnostics) Append(in ...Diagnostic)
+func (d Diagnostics) HasError() bool
+func (d Diagnostics) toTfprotov6Diagnostics() []*tfprotov6.Diagnostic
+```
+
+With an example implementation:
+
+```go
+// ErrorDiagnostic is a generic diagnostic with error severity.
+type ErrorDiagnostic struct {
+  Summary string
+  Detail  string
+}
+
+func (d ErrorDiagnostic) Severity() Severity { return SeverityError }
+func (d ErrorDiagnostic) Summary() string { return d.Summary }
+func (d ErrorDiagnostic) Detail() string { return d.Detail }
+func (d ErrorDiagnostic) Equal(other Diagnostic) bool {
+  other, ok := other.(ErrorDiagnostic)
+
+  if !ok {
+    return false
+  }
+
+  return other.Summary() == d.Summary() && other.Detail() == d.Detail()
+}
+
+// ValidationErrorDiagnostic represents an error during validation.
+type ValidationErrorDiagnostic struct {
+  ErrorDiagnostic
+}
+
+func (d ValidationErrorDiagnostic) Equal(other Diagnostic) bool {
+  other, ok := other.(ValidationErrorDiagnostic)
+
+  if !ok {
+    return false
+  }
+
+  return other.Summary() == d.Summary() && other.Detail() == d.Detail()
+}
+```
+
+It then becomes possible to distinguish between diagnostic types (in this case the "source" of a diagnostic) while still supporting additional functionality defined by the interface. These extended diagnostics can also implement other framework-defined capabilities, such as implementing the example `Log()` method in this situation.
```

Comment:
I think the use of "source" is throwing me, because I keep expecting it to be able to work like a stacktrace, helping me find the line a diagnostic was returned from. But I think the comment about distinguishing between the type of diagnostics is really illuminating.